CRM/ManualDirectDebit/Form/SignUp.php

```diff
@@ -0,0 +1,48 @@
+<?php
+
+use CRM_ManualDirectDebit_ExtensionUtil as E;
+
+/**
+ * Form Manual Direct Debit Sign up Form
+ *
+ */
+class CRM_ManualDirectDebit_Form_SignUp extends CRM_Core_Form {
+
+  private $contributionId;
+
+  public function preProcess() {
+    parent::preProcess();
+    $this->contributionId = CRM_Utils_Request::retrieveValue('contribution_id', 'String', NULL);
+    if ($this->contributionId == NULL) {
+      CRM_Utils_System::redirect('/');
+    }
+    CRM_Utils_System::setTitle(E::ts('Direct Debit Sign up'));
+  }
+
+  public function buildQuickForm() {
+    $errorMessage = NULL;
+    $contribution = civicrm_api3('Contribution', 'get', [
+      'sequential' => 1,
+      'id' => $this->contributionId,
+      'contribution_status_id' => "Pending",
+    ]);
+
+    if ($contribution['count'] == 0) {
```

Comment:
What if I use a contribution ID that does not exist? For example, I could fabricate the URL like `civicrm/direct_debit/signup?contribution_id=ABC123`. In that case, since `count` will be 0, it'll say the the contribution is already paid. Is that a problem? Do we need to handle that case separately?

eg.
```php
$contribution = $this->getContribution();

if (empty($contribution)) {
  $errorMessage = E::ts("Contribution with ID {$this->contributionId} does not exist... etc.");
  $this->assign('errorMessage', $errorMessage);

  return;
}

if (!$this->isContributionPending($contribution)) {
  $errorMessage = E::ts('This invoice is already paid. Please contact the administrator for the correct link to setup direct debit.');
} else if (is_null($contribution['value']['contribution_recur_id'])) {
  $errorMessage = E::ts('This invoice is not part of a payment plan. Please contact the administrator for the correct link to set up a direct debit.');
}
...
```

Reply:
@MiyaNoctem  this has already been handled by preProcess() as it will be redirected to the home page if contribution id is not passed in the URL and the if the contribution is already paid the count will always be 0 as we passed parameter to check contribution_status_id when calling CiviCRM API. 

```

 $contribution = civicrm_api3('Contribution', 'get', [
      'sequential' => 1,
      'id' => $this->contributionId,
      'contribution_status_id' => "Pending",
    ]);
```

Reply:
But I would be passing a contribution ID. Just an ID that does not exist in the database. If I do that, it'll still say the contributions is paid, which is not accurate, the reality is that it does not exist. Not sure if that is a problem or not, though, it might be fine. For example, say there is only one contribution in the database with ID = 1, and I go to `civicrm/direct_debit/signup?contribution_id=999999999999`. I'tll say contribution with ID 999999999999 is already paid.

Reply:
@MiyaNoctem  I don't think it would be a problem as the result will be 0. The user need to contact the admin.